core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,541 @@
+                    state_session_ids_by_version[str(value['version'])] = {}
+                if value['state_name'] not in state_hit_counts_by_version[
+                        str(value['version'])]:
+                    state_hit_counts_by_version[str(value['version'])][value[
+                        'state_name']] = {
+                            'total_hit_count': 0,
+                            'first_hit_count': 0
+                        }
+                if value['state_name'] not in state_session_ids_by_version[
+                        str(value['version'])]:
+                    state_session_ids_by_version[str(value['version'])][
+                        value['state_name']] = set()
+                state_hit_counts_by_version[str(value['version'])][value[
+                    'state_name']]['total_hit_count'] += 1
+                if value['session_id'] not in state_session_ids_by_version[
+                        str(value['version'])][value['state_name']]:
+                    state_hit_counts_by_version[str(value['version'])][value[
+                        'state_name']]['first_hit_count'] += 1
+                    state_session_ids_by_version[str(value['version'])][value[
+                        'state_name']].add(value['session_id'])
+
+                if str(value['version']) not in (
+                        session_id_state_hit_event_mapping):
+                    session_id_state_hit_event_mapping[
+                        str(value['version'])] = {}
+                if value['session_id'] not in (
+                        session_id_state_hit_event_mapping[str(value[
+                            'version'])]):
+                    session_id_state_hit_event_mapping[str(value['version'])][
+                        value['session_id']] = []
+                session_id_state_hit_event_mapping[str(value['version'])][
+                    value['session_id']].append(value)
+            elif value['event_type'] == (
+                    GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS):
+                if str(value['version']) not in state_answer_counts_by_version:
+                    state_answer_counts_by_version[str(value['version'])] = {}
+                state_answer_counts_by_version[str(value['version'])][
+                    value['state_name']] = {
+                        'total_answers_count': value['total_answers_count'],
+                        'useful_feedback_count': value['useful_feedback_count']
+                    }
+
+        # This dict will store the cumulative values of state stats from version
+        # 1 till the latest version of an exploration.
+        state_stats_mapping = {}
+        # This list will contain all the ExplorationStats domain objects for
+        # each version.
+        exploration_stats_list = []
+        for version in version_numbers:
+            state_hit_counts_for_this_version = {}
+            if str(version) in state_hit_counts_by_version:
```

Comment:
If you use collections.defaultdict then you shouldn't need this and other if clauses.

Reply:
Yep. Removed.